python-package/lightgbm/dask.py

```diff
@@ -358,11 +394,11 @@ def _predict(model, data, raw_score=False, pred_proba=False, pred_leaf=False, pr
 
     Returns
     -------
-    predicted_result : dask array of shape = [n_samples] or shape = [n_samples, n_classes]
+    predicted_result : dask Array of shape = [n_samples] or shape = [n_samples, n_classes]
         The predicted values.
-    X_leaves : dask array of shape = [n_samples, n_trees] or shape = [n_samples, n_trees * n_classes]
+    X_leaves : dask Array of shape = [n_samples, n_trees] or shape = [n_samples, n_trees * n_classes]
         If ``pred_leaf=True``, the predicted leaf of every tree for each sample.
-    X_SHAP_values : dask array of shape = [n_samples, n_features + 1] or shape = [n_samples, (n_features + 1) * n_classes] or list with n_classes length of such objects
+    X_SHAP_values : dask Array of shape = [n_samples, n_features + 1] or shape = [n_samples, (n_features + 1) * n_classes] or list with n_classes length of such objects
```

Comment:
> ... or list with n_classes length of such objects

Is it supported?

Reply:
it looks like `dask-lightgbm` probably just copied that from https://github.com/microsoft/LightGBM/blob/c4b51317f8f722f60105335e5c799877a7dd80bd/python-package/lightgbm/sklearn.py#L691.

How does the sklearn model object produce an output like that?

Reply:
Yes, I simply copied this line.

It comes from here:
https://github.com/microsoft/LightGBM/blob/c4b51317f8f722f60105335e5c799877a7dd80bd/python-package/lightgbm/basic.py#L678
https://github.com/microsoft/LightGBM/blob/c4b51317f8f722f60105335e5c799877a7dd80bd/python-package/lightgbm/basic.py#L805-L849
Contributed in #3000.

Reply:
From that PR, it looks like this is specific to sparse inputs? I'll have to test that with a sparse Dask Array. Can try that later tonight and report back.

Reply:
Here are corresponding tests you can refer to (some were added later in separate RPs):
https://github.com/microsoft/LightGBM/blob/42d1633aebe124821cff42c728a42551db715168/tests/python_package_test/test_engine.py#L1025-L1123

Reply:
Thanks for the links!

Ok, I tested some stuff tonight. I found that right now, if you provide a Dask Array whose pieces are CSC matrices to `.predict(pred_contrib = TRUE)`, you get back a Dask Array whose individual partitions are also CSC matrices. Calling `.compute()` on that Dask array returns a `scipy.sparse.coo.coo_matrix`.

That returned matrix does have the expected dimensions. For an input of shape (100, 2), for `n_classes=3`, you get back a matrix of shape `(100, 9)`.

So I think the "or list with n_classes..." part should be removed from this docstring. What do you think?

```python
import dask.array as da
import numpy as np
from dask.distributed import Client, LocalCluster, wait
from lightgbm.dask import DaskLGBMClassifier
from lightgbm.sklearn import LGBMClassifier
from scipy.sparse import csc_matrix
from sklearn.datasets import make_blobs

n_workers = 3
cluster = LocalCluster(n_workers=n_workers)
client = Client(cluster)
client.wait_for_workers(n_workers)

print(f"View the dashboard: {cluster.dashboard_link}")

chunk_size=50
X, y = make_blobs(n_samples=100, centers=3, random_state=42)
rnd = np.random.RandomState(42)
dX = da.from_array(X, chunks=(chunk_size, X.shape[1])).map_blocks(csc_matrix)
dy = da.from_array(y, chunks=chunk_size)

#  Dask predict() creates a Dask Array, whose `.compute()` is a 'scipy.sparse.coo.coo_matrix'
dask_clf = DaskLGBMClassifier(n_estimators=5, num_leaves=2, tree_learner="data")
dask_clf.fit(client, dX, dY)

preds = dask_clf.predict(dX, pred_contrib=True)
preds_computed = preds.compute()

print(type(preds), type(preds_computed), f"{dask_clf.n_classes_} classes, {dX.shape[1]} features")
print("---")
print(dX.partitions[0].compute())
print("---")
preds.compute().shape

# local predict() returns a list of 3 CSC matrices
X = dX.compute()
y = dy.compute()

local_clf = LGBMClassifier()
local_clf.fit(X=dX.compute(), y=y)
local_preds = local_clf.predict(dX.compute().tocsc(), pred_contrib=True)

print(local_clf.n_classes_, type(local_preds))
print("---")
print(local_preds)
```

Reply:
I made this change in https://github.com/microsoft/LightGBM/pull/3866/commits/7b3a0e569df5c5db0f6454e9c85f313cec35889c, but won't merge until you have a chance to comment

Reply:
I'm sorry, I don't have an easy access to Dask environment where I can run tests with LightGBM due to that Windows isn't supported right now. So, I'm OK with whatever you decide. By my initial comment I just wanted to bring to your attention that there is a special case of return type for sparse matrices in the underlying API.

I just wonder where Dask merges `n_classes` sparse matrices into 1 like here
https://github.com/microsoft/LightGBM/blob/42d1633aebe124821cff42c728a42551db715168/tests/python_package_test/test_engine.py#L1083-L1086

Because 
> ```#  Dask predict() creates a Dask Array, whose `.compute()` is a 'scipy.sparse.coo.coo_matrix'```

`coo_matrix` should be casted to `csr` one here
https://github.com/microsoft/LightGBM/blob/c4b51317f8f722f60105335e5c799877a7dd80bd/python-package/lightgbm/basic.py#L721-L727
and inner reshaping should not be happen:
https://github.com/microsoft/LightGBM/blob/c4b51317f8f722f60105335e5c799877a7dd80bd/python-package/lightgbm/basic.py#L730-L736

Reply:
thanks for pointing it out! I'm not sure why that's happening. I think having a return type of `List[array]` is going to be really challenging for Dask, but either way we can figure it out in a separate issue.

Recall that for this PR, the goal is

> This PR attempts to document the current state of the package, and isn't a proposal about the final state for release 3.2.0

And the current state is that if you pass a Dask Array whose parts are CSC matrices into `.predict(pred_contrib=True)`, you will get back a Dask Array.

I added a feature request to resolve the inconsistency. I think it's going to be really difficult to do, but can be figured out at a later time and doesn't need to block this PR: https://github.com/microsoft/LightGBM/issues/3881

Reply:
Thank you! Agree.